### PR TITLE
[YLP 324] pump and cgm information dbl settings

### PR DESCRIPTION
**Pull request by Nas84**

### CODE REVIEW

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
Quick review (not tested yet)
Missing translations in locales/en/translations.json

File: packages/viz/src/components/settings/CgmTable.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          <tr><td>{t('Transmitter ID')}</td><td>{this.props.cgm.transmitterId}</td></tr>
+          <tr><td>{t('Transmitter end of life')}</td><td>{cgmEndOfLife}</td></tr>
+        </tbody>
+      </table>
+    );
+  }
+
+  formatDate(value) {
+    if (_.isEmpty(value))
+    {
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/CgmTable.js:38]:
```suggestion
    if (_.isEmpty(value)) {
```
We should kept the Java/Javascript style I think here.

File: packages/viz/src/components/settings/PumpTable.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+          <tr><td>{t('Pump Version')}</td><td>{this.props.pump.swVersion}</td></tr>
+          <tr><td>{t('Cartridge expiration date')}</td><td>{pumpExpirationDate}</td></tr>
+        </tbody>
+      </table>
+    );
+  }
+
+  formatDate(value) {
+    if (_.isEmpty(value))
+    {
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/PumpTable.js:35]:
```suggestion
    if (_.isEmpty(value)) {
```
We should kept the Java/Javascript style I think here.

File: packages/viz/src/components/settings/CgmTable.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import i18next from 'i18next';
+import _ from 'lodash';
+import styles from './Diabeloop.css';
+import * as datetime from '../../utils/datetime';
+
+const t = i18next.t.bind(i18next);
+
+class CgmTable extends React.Component {
+  constructor(props) {
+      super(props);
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/CgmTable.js:12]:
```suggestion
    super(props);

```

File: packages/viz/src/components/settings/PumpTable.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import _ from 'lodash';
+import i18next from 'i18next';
+import styles from './Diabeloop.css';
+import * as datetime from '../../utils/datetime';
+
+const t = i18next.t.bind(i18next);
+
+class PumpTable extends React.Component {
+  constructor(props) {
+      super(props);
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/PumpTable.js:12]:
```suggestion
    super(props);

```

File: packages/viz/src/components/settings/PumpTable.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  render() {
+    const pumpExpirationDate = this.formatDate(this.props.pump.expirationDate);
+    
+    return (
+      <table className={styles.settingsTable}>
+        <caption className={styles.bdlgSettingsHeader}>
+          {t('Pump')}
+        </caption>
+        <tbody>
+          <tr><td>{t('Manufacturer')}</td><td>{this.props.pump.manufacturer}</td></tr>
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/PumpTable.js:24]: Cleaner to use at the beginning of the function:
```
const  { pump} = this.props;
```

File: packages/viz/src/components/settings/PumpTable.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import moment from 'moment-timezone';
+import _ from 'lodash';
+import i18next from 'i18next';
+import styles from './Diabeloop.css';
+import * as datetime from '../../utils/datetime';
+
+const t = i18next.t.bind(i18next);
+
+class PumpTable extends React.Component {
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/PumpTable.js:10]: No state in this component, a function may be used instead

File: packages/viz/src/components/settings/CgmTable.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    );
+  }
+
+  formatDate(value) {
+    if (_.isEmpty(value))
+    {
+      return null;
+    }
+
+    return datetime.formatLocalizedFromUTC(value, this.props.timePrefs, 'MMM D, YYYY');
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/CgmTable.js:42]: Date format should be translated (key already exists)

File: packages/viz/src/components/settings/PumpTable.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    );
+  }
+
+  formatDate(value) {
+    if (_.isEmpty(value))
+    {
+      return null;
+    }
+
+    return datetime.formatLocalizedFromUTC(value, this.props.timePrefs, 'MMM D, YYYY');
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/PumpTable.js:39]: Date format should be translated (key already exists)

File: packages/viz/src/components/settings/CgmTable.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import moment from 'moment-timezone';
+import i18next from 'i18next';
+import _ from 'lodash';
+import styles from './Diabeloop.css';
+import * as datetime from '../../utils/datetime';
+
+const t = i18next.t.bind(i18next);
+
+class CgmTable extends React.Component {
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/CgmTable.js:10]: No state in this component, a function may be used instead

File: packages/viz/src/components/settings/CgmTable.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+const t = i18next.t.bind(i18next);
+
+class CgmTable extends React.Component {
+  constructor(props) {
+      super(props);
+  }
+
+  render() {
+    const cgmEndOfLife = this.formatDate(this.props.cgm.endOfLifeTransmitterDate);
```
Comment (yvon-dblg) [inline packages/viz/src/components/settings/CgmTable.js:16]: Cleaner to use at the beginning of the function:
```
const  { cgm } = this.props;
```

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
- Should display "N/A" in case of missing information.
- Missing PropTypes validations
- Device information table has disappear !
- The "DBL" (`t('Device')`) alone seems a little bit odd now, we should see with @MarineDBL if it need an adjustment.

**Comment by Nas84** [comment]
- missing information will be marked as "-" hyphen 
- As discussed, i added the validation and the default value.
- Device table is back 😄  it was a spec misunderstand

thanks for your help it is great 😉

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
Looks good !
If some info are missing, we may have a crash
Here a proposition to avoid that:
```patch
diff --git a/packages/viz/src/components/settings/CgmTable.js b/packages/viz/src/components/settings/CgmTable.js
index 93f914900..0fea8a817 100644
--- a/packages/viz/src/components/settings/CgmTable.js
+++ b/packages/viz/src/components/settings/CgmTable.js
@@ -15,9 +15,14 @@ class CgmTable extends React.Component {
 
   render() {
     const { cgm } = this.props;
+
+    if (cgm === null) {
+      return null;
+    }
+
     const cgmEndOfLife = this.formatDate(cgm.endOfLifeTransmitterDate);
     const cgmSensorExpirationDate = this.formatDate(cgm.expirationDate);
-  
+
     return (
       <table className={styles.cgmTable}>
         <caption className={styles.bdlgSettingsHeader}>
@@ -69,7 +74,7 @@ CgmTable.propTypes = {
   timePrefs: PropTypes.shape({
     timezoneAware: PropTypes.bool.isRequired,
     timezoneName: PropTypes.string,
-  }).isRequired
+  }).isRequired,
 };
 
 export default CgmTable;
diff --git a/packages/viz/src/components/settings/Diabeloop.js b/packages/viz/src/components/settings/Diabeloop.js
index ecc4b5529..c7edfbb1c 100644
--- a/packages/viz/src/components/settings/Diabeloop.js
+++ b/packages/viz/src/components/settings/Diabeloop.js
@@ -105,10 +105,6 @@ const Diabeloop = (props) => {
   const cgm = _.get(pumpSettings, 'payload.cgm', null);
   const history = _.sortBy(_.cloneDeep(_.get(pumpSettings, 'payload.history', null)), ['changeDate']);
 
-  if (parameters === null || device === null) {
-    return null;
-  }
-
   const parametersByLevel = dblData.getParametersByLevel(parameters);
 
   return (
@@ -129,9 +125,9 @@ const Diabeloop = (props) => {
       <div className={styles.settingsContainer}>
         <div className={styles.categoryContainer}>
           <div className={styles.categoryTitle}>{t('Device')}</div>
-          <TerminalTable device={device}> </TerminalTable>
-          <PumpTable pump={pump} timePrefs={timePrefs}> </PumpTable>
-          <CgmTable cgm={cgm} timePrefs={timePrefs}> </CgmTable>
+          <TerminalTable device={device} />
+          <PumpTable pump={pump} timePrefs={timePrefs} />
+          <CgmTable cgm={cgm} timePrefs={timePrefs} />
         </div>
         {renderDiabeloopParameters(parametersByLevel)}
         <div className={styles.categoryContainer}>
@@ -164,6 +160,7 @@ Diabeloop.propTypes = {
   deviceKey: PropTypes.oneOf(['diabeloop']).isRequired,
   pumpSettings: PropTypes.shape({
     deviceId: PropTypes.string.isRequired,
+    deviceTime: PropTypes.string.isRequired,
     payload: PropTypes.shape({
       device: PropTypes.shape({
         deviceId: PropTypes.string.isRequired,
@@ -201,5 +198,6 @@ Diabeloop.propTypes = {
     timezoneName: PropTypes.string,
   }).isRequired,
   user: PropTypes.object.isRequired,
+  handleClickHistory: PropTypes.func.isRequired,
 };
 export default Diabeloop;
diff --git a/packages/viz/src/components/settings/PumpTable.js b/packages/viz/src/components/settings/PumpTable.js
index bc2d3d8b8..a4800a041 100644
--- a/packages/viz/src/components/settings/PumpTable.js
+++ b/packages/viz/src/components/settings/PumpTable.js
@@ -14,9 +14,13 @@ class PumpTable extends React.Component {
   }
 
   render() {
-    const { pump } = this.props; 
+    const { pump } = this.props;
     const pumpExpirationDate = this.formatDate(pump.expirationDate);
-    
+
+    if (pump === null) {
+      return null;
+    }
+
     return (
       <table className={styles.pumpTable}>
         <caption className={styles.bdlgSettingsHeader}>
diff --git a/packages/viz/src/components/settings/TerminalTable.js b/packages/viz/src/components/settings/TerminalTable.js
index ea13fa30c..a999c203b 100644
--- a/packages/viz/src/components/settings/TerminalTable.js
+++ b/packages/viz/src/components/settings/TerminalTable.js
@@ -12,7 +12,11 @@ class TerminalTable extends React.Component {
   }
 
   render() {
-    const { device } = this.props; 
+    const { device } = this.props;
+
+    if (device === null) {
+      return null;
+    }
 
     return (
       <table className={styles.deviceTable}>
@@ -48,7 +52,7 @@ TerminalTable.propTypes = {
     name: PropTypes.string.isRequired,
     manufacturer: PropTypes.string.isRequired,
     swVersion: PropTypes.string.isRequired,
-  }).isRequired
-}; 
+  }).isRequired,
+};
 
 export default TerminalTable;
diff --git a/packages/viz/src/utils/settings/diabeloopData.js b/packages/viz/src/utils/settings/diabeloopData.js
index f8b1b0c29..bd39a3b0b 100644
--- a/packages/viz/src/utils/settings/diabeloopData.js
+++ b/packages/viz/src/utils/settings/diabeloopData.js
@@ -30,20 +30,22 @@ export function getParametersByLevel(parameters) {
   const mapParams = new Map();
 
   // eslint-disable-next-line lodash/prefer-lodash-method
-  parameters.forEach((parameter) => {
-    if (!mapParams.has(parameter.level)) {
-      mapParams.set(parameter.level, []);
-    }
-
-    const value = formatParameterValue(parameter.value, parameter.unit);
-    const param = {
-      name: t(`params:::${parameter.name}`),
-      value,
-      unit: parameter.unit,
-      level: parameter.level,
-    };
-    mapParams.get(parameter.level).push(param);
-  });
+  if (Array.isArray(parameters)) {
+    parameters.forEach((parameter) => {
+      if (!mapParams.has(parameter.level)) {
+        mapParams.set(parameter.level, []);
+      }
+
+      const value = formatParameterValue(parameter.value, parameter.unit);
+      const param = {
+        name: t(`params:::${parameter.name}`),
+        value,
+        unit: parameter.unit,
+        level: parameter.level,
+      };
+      mapParams.get(parameter.level).push(param);
+    });
+  }
 
   return mapParams;
 }

```

**Review by yvon-dblg — APPROVED** [review]
LGTM

**Pull request merged**